### PR TITLE
[CARBONDATA-3641] Refactory data loading for partition table

### PR DESCRIPTION
**Pull request by QiangCai**
[Background]
1. only implemented commit algorithm version 1
2. generated too many segment files during loading
3. generated too many small data files and index files

[Modification]
1.  implemented carbon commit algorithm, avoid to move data file and index files
2.  generate the final segment file directly
3.   optimize global_sort to avoid small files issue
4. normal columns support complex data type

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1307/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1298/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1318/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1306/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1316/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1329/

**Comment by QiangCai** [comment]
please retest this

**Comment by QiangCai** [comment]
retest this please

File: core/src/main/java/org/apache/carbondata/core/metadata/SegmentFileStore.java [hunk 1]
```diff
@@ -1228,7 +1228,7 @@ public static void removeTempFolder(Map<String, FolderDetails> locationMap, Stri
       locationMap = new HashMap<>();
     }
 
-    SegmentFile merge(SegmentFile mapper) {
+    public SegmentFile merge(SegmentFile mapper) {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/metadata/SegmentFileStore.java]: name `mapper` is not good, can you rename it?

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 2]
```diff
@@ -2730,39 +2734,66 @@ public static String encodeToString(byte[] bytes) throws UnsupportedEncodingExce
     return Base64.decodeBase64(objectString.getBytes(CarbonCommonConstants.DEFAULT_CHARSET));
   }
 
+  public static void copyCarbonDataFileToCarbonStorePath(String localFilePath,
+      String carbonDataDirectoryPath, long fileSizeInBytes,
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]:
```suggestion
      String targetPath, long fileSizeInBytes,
```

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 3]
```diff
@@ -2730,39 +2734,66 @@ public static String encodeToString(byte[] bytes) throws UnsupportedEncodingExce
...
+      OutputFilesInfoHolder outputFilesInfoHolder) throws CarbonDataWriterException {
+    if (carbonDataDirectoryPath.endsWith(".tmp") && localFilePath
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: move localFilePath to next line

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 4]
```diff
@@ -2730,39 +2734,66 @@ public static String encodeToString(byte[] bytes) throws UnsupportedEncodingExce
...
+      outputFilesInfoHolder.addToOutputFiles(carbonDataDirectoryPath + localFilePath
+          .substring(localFilePath.lastIndexOf(File.separator)) + ":" + targetSize);
+    }
+  }
+
   /**
    * This method will copy the given file to carbon store location
    *
    * @param localFilePath local file name with full path
    * @throws CarbonDataWriterException
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: add @return description

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 5]
```diff
@@ -2730,39 +2734,66 @@ public static String encodeToString(byte[] bytes) throws UnsupportedEncodingExce
...
-          getMaxOfBlockAndFileSize(fileSizeInBytes, localCarbonFile.getSize()));
+          getMaxOfBlockAndFileSize(fileSizeInBytes, localFileSize));
       CarbonFile targetCarbonFile = FileFactory.getCarbonFile(carbonFilePath);
       // the size of carbon file must be greater than 0
       // and the same as the size of local carbon file
-      if (targetCarbonFile.getSize() == 0L ||
-          (targetCarbonFile.getSize() != localCarbonFile.getSize())) {
+      targetSize = targetCarbonFile.getSize();
+      if (targetSize == 0L ||
+          (targetSize != localFileSize)) {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]:
```suggestion
          targetSize != localFileSize) {
```

File: core/src/main/java/org/apache/carbondata/core/util/comparator/BigDecimalSerializableComparator.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    if (key1 == null && key2 == null) {
+      return 0;
+    } else if (key1 == null) {
+      return -1;
+    } else if (key2 == null) {
+      return 1;
+    }
+    return ((BigDecimal) key1).compareTo((BigDecimal) key2);
+  }
+}
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/util/comparator/BigDecimalSerializableComparator.java]: add new line at the end

File: core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java [hunk 7]
```diff
@@ -116,6 +119,66 @@ private String mergeCarbonIndexFilesOfSegment(String segmentId,
     return null;
   }
 
+  public SegmentFileStore.FolderDetails mergeCarbonIndexFilesOfSegment(String segmentId,
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java]: add description for the func

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonOutputCommitter.java [hunk 8]
```diff
@@ -222,6 +215,91 @@ public void commitJob(JobContext context) throws IOException {
     }
   }
 
+  private void commitJobFinal(JobContext context, CarbonLoadModel loadModel,
```
Comment (jackylk) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonOutputCommitter.java]: add description

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonOutputCommitter.java [hunk 9]
```diff
@@ -222,6 +215,91 @@ public void commitJob(JobContext context) throws IOException {
...
+    if (updateTime != null) {
+      CarbonUpdateUtil.updateTableMetadataStatus(segmentSet, carbonTable, updateTime, true,
+          segmentDeleteList);
+    } else if (uniqueId != null) {
+      CarbonUpdateUtil.updateTableMetadataStatus(segmentSet, carbonTable, uniqueId, true,
+          segmentDeleteList);
+    }
+  }
+
+  private void commitJobForPartition(JobContext context, boolean overwriteSet,
```
Comment (jackylk) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonOutputCommitter.java]: add description

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableOutputFormat.java [hunk 10]
```diff
@@ -481,10 +485,42 @@ public void close(TaskAttemptContext taskAttemptContext) throws InterruptedExcep
           // clean up the folders and files created locally for data load operation
           TableProcessingOperations.deleteLocalDataLoadFolderLocation(loadModel, false, false);
         }
+        OutputFilesInfoHolder outputFilesInfoHolder = loadModel.getOutputFilesInfoHolder();
+        if (null != outputFilesInfoHolder) {
+          //TODO: fix to sum
```
Comment (jackylk) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableOutputFormat.java]: Is this TODO required?

File: integration/spark-common/src/main/scala/org/apache/spark/rdd/CarbonMergeFilesRDD.scala [hunk 11]
```diff
@@ -181,17 +264,39 @@ class CarbonMergeFilesRDD(
...
-      if (isHivePartitionedTable) {
+      var segmentFile: SegmentFileStore.SegmentFile = null
+      var indexSize: String = ""
+      if (isHivePartitionedTable && partitionInfo.isEmpty) {
         CarbonLoaderUtil
           .mergeIndexFilesInPartitionedSegment(carbonTable, split.segmentId,
             segmentFileNameToSegmentIdMap.get(split.segmentId), split.partitionPath)
+      } else if (isHivePartitionedTable && !partitionInfo.isEmpty) {
+        val folderDetails = CarbonLoaderUtil
+          .mergeIndexFilesInPartitionedTempSegment(carbonTable,
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/spark/rdd/CarbonMergeFilesRDD.scala]: move `carbonTable` to next line

File: integration/spark2/src/main/scala/org/apache/spark/sql/events/MergeIndexEventListener.scala [hunk 12]
```diff
@@ -118,6 +147,8 @@ class MergeIndexEventListener extends OperationEventListener with Logging {
                 carbonTable = carbonMainTable,
                 mergeIndexProperty = true,
                 readFileFooterFromCarbonDataFile = true)
+              LOGGER.info("Total time taken for merge index "
+                          + (System.currentTimeMillis() - startTime))
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/events/MergeIndexEventListener.scala]:
```suggestion
                          + (System.currentTimeMillis() - startTime) + "ms")
```

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala [hunk 13]
```diff
@@ -206,16 +215,193 @@ with Serializable {
...
+      }
+      super
+        .commitJob(jobContext, newTaskCommits)
+    } else {
+      super
+        .commitJob(jobContext, taskCommits)
+    }
+  }
+
+  override def commitTask(
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala]: add description

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala [hunk 14]
```diff
@@ -206,16 +215,193 @@ with Serializable {
...
   extends SQLHadoopMapReduceCommitProtocol(jobId, path, isAppend) {
+
+  override def setupTask(taskContext: TaskAttemptContext): Unit = {
+    if (isCarbonDataFlow(taskContext.getConfiguration)) {
+      ThreadLocalSessionInfo.setConfigurationToCurrentThread(taskContext.getConfiguration)
+    }
+    super.setupTask(taskContext)
+  }
+
+  override def commitJob(jobContext: JobContext,
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala]: add description

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala [hunk 15]
```diff
@@ -206,16 +215,193 @@ with Serializable {
...
-    if (carbonFlow != null) {
+    if (isCarbonFileFlow(taskContext.getConfiguration) ||
+        isCarbonDataFlow(taskContext.getConfiguration)) {
       super.newTaskTempFile(taskContext, Some(absoluteDir), ext)
     } else {
       super.newTaskTempFileAbsPath(taskContext, absoluteDir, ext)
     }
   }
+
+  override def newTaskTempFile(taskContext: TaskAttemptContext,
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala]: add description,
move `taskContext` to next line

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala [hunk 16]
```diff
@@ -432,4 +541,77 @@ private class CarbonOutputWriter(path: String,
...
+      (col, null)
+      // we should replace back the special string with empty value.
+    } else if (value.equals(CarbonCommonConstants.MEMBER_DEFAULT_VAL)) {
+      (col, "")
+    } else {
+      (col, value)
+    }
+  }
+
+  def updatePartitions(
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala]: add descripiton

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1332/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1342/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1355/

File: core/src/main/java/org/apache/carbondata/core/metadata/SegmentFileStore.java (hunk 1 above)
Comment (QiangCai) [reply core/src/main/java/org/apache/carbondata/core/metadata/SegmentFileStore.java]: done

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java (hunk 2 above)
Comment (QiangCai) [reply core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: done

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java (hunk 3 above)
Comment (QiangCai) [reply core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: done

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java (hunk 4 above)
Comment (QiangCai) [reply core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: done

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java (hunk 5 above)
Comment (QiangCai) [reply core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: done

File: core/src/main/java/org/apache/carbondata/core/util/comparator/BigDecimalSerializableComparator.java (hunk 6 above)
Comment (QiangCai) [reply core/src/main/java/org/apache/carbondata/core/util/comparator/BigDecimalSerializableComparator.java]: done

File: core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java (hunk 7 above)
Comment (QiangCai) [reply core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java]: done

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonOutputCommitter.java (hunk 8 above)
Comment (QiangCai) [reply hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonOutputCommitter.java]: done

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonOutputCommitter.java (hunk 9 above)
Comment (QiangCai) [reply hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonOutputCommitter.java]: done

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableOutputFormat.java (hunk 10 above)
Comment (QiangCai) [reply hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableOutputFormat.java]: no need

File: integration/spark-common/src/main/scala/org/apache/spark/rdd/CarbonMergeFilesRDD.scala (hunk 11 above)
Comment (QiangCai) [reply integration/spark-common/src/main/scala/org/apache/spark/rdd/CarbonMergeFilesRDD.scala]: done

File: integration/spark2/src/main/scala/org/apache/spark/sql/events/MergeIndexEventListener.scala (hunk 12 above)
Comment (QiangCai) [reply integration/spark2/src/main/scala/org/apache/spark/sql/events/MergeIndexEventListener.scala]: accepted

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala (hunk 13 above)
Comment (QiangCai) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala]: done

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala (hunk 14 above)
Comment (QiangCai) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala]: done

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala (hunk 15 above)
Comment (QiangCai) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala]: done

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala (hunk 16 above)
Comment (QiangCai) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/SparkCarbonTableFormat.scala]: done

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1342/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1352/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1365/

**Comment by jackylk** [comment]
LGTM
spark 2.3 integration CI has passed

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1368/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1347/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1357/

**Comment by brijoobopanna** [comment]
@QiangCai please check if this PR has induced any CI issues to open source

**Comment by akashrn5** [comment]
@QiangCai i think spark2.2 CI is continously failing with this PR, can you please have a look at it?

**Pull request closed without merging**